New Pick Parenting for Ray Picks and Parabola Picks

sabrina-shanman opened the pull request:
https://highfidelity.manuscript.com/f/cases/19083

This PR is a follow-up to https://github.com/highfidelity/hifi/pull/13873, which added parenting to just Collision Picks.

This PR is not a blocker for https://github.com/highfidelity/hifi/pull/13937

To recap: 
- A pick calculates spatial information about the world on the C++ side. Picks are immutable to allow efficient use by scripts.
- A collision pick calculates collision of a volume in the physics engine
- Ray picks and parabola picks follow along a path until they hit something, then share this information
- Pick parenting allows the pick's position, orientation, and scale (where applicable) to be calculated relative to another object, without creating a new pick.

This PR makes Ray Picks and Parabola Picks use the same parenting system as Collision Picks currently do. This adds some new functionality for Ray and Parabola Picks, and cleans up some redundant code.

In summary, the new parenting system exposes the following on the javascript API side:
- joint - A string representing either a joint on My Avatar to parent the pick to, My Avatar's head (which is currently not a real joint), or the mouse. This option currently exists for Ray and Parabola Picks. It has been included for consistency, and may be deprecated in the future.
- parentID - Defines a spatially nestable (avatar, entity, or 3d overlay), or another pick to parent the pick to. This is a new feature for Ray and Parabola Picks, and currently exists for Collision Picks.
- parentJointIndex - Goes along with parentID. Defines, optionally, a numeric joint ID on the spatially nestable parent to be parented to

A description of how pick parenting is currently implemented has also been added as documentation via markdeep. You can view that [here](https://github.com/sabrina-shanman/hifi/blob/stt_parenting2/docs/interface/raypick/pick-parenting.md.html)

There are a few aspects of this PR that I have questions on, which I will describe in the [next post](https://github.com/highfidelity/hifi/pull/14000#issuecomment-421449174).

## TEST PLAN
- Run the tests in these folders:
  - https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer/parabola
  - https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pointer/laser
  - https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/interaction/pick/collision
    - Can ignore the capsule_on_server test
- Run these tests in this PR and master and confirm behavior is the same:
  - Compare the shape and size of the parabola when about to teleport at different scales of the avatar and at different controller orientations. In addition to the trajectory of the parabola, the thickness of the parabola, and the teleportation visualization at the end of the parabola, should look the same size in this PR as in master.
  - At different scales of the avatar, compare the lack of ability of the avatar to teleport into tight spaces, i.e. when the ceiling is too low or when the region to teleport into is very narrow.
  - Compare the appearance of the laser pointer when pointing at terrain at different distances of the avatar, while at different scales of the avatar. Also compare the appearance of the laser pointer when grabbing objects. The thickness of the laser and the spherical point at the end should look the same size in this PR as in master.

sabrina-shanman commented:
### Caveats/Questions
- In this PR, Stylus Picks still have their own system where they can only be parented to the left/right hand. Is this sufficient?
- This PR includes a change in the javascript API that could *potentially* break scripts, but given the change improves the API, the API is not widely used, and the use case for the removed behavior is questionable, I would be curious to hear your thoughts. In the old behavior, the Parabola Pick had an option called "scaleWithAvatar," which always scaled the parabola to the avatar regardless of what the parabola was parented to. In the new behavior, the Parabola Pick has an option called "scaleWithParent," with "scaleWithAvatar" as an alias. The new behavior scales the Parabola Pick according to its current parent.
- This PR also introduces a new feature called "rotateAccelerationWithParent," which is NOT an alias to the currently existing "rotateAccelerationWithAvatar." It has been added because it seemed odd to have a feature specific to My Avatar, however is it needed? It causes the parabola to bend "down," where "down" is relative to the orientation of the parent.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13010/HighFidelity-Beta-PR14000-237022d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13010/HighFidelity-Beta-Interface-PR14000-237022d.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13010/HighFidelity-Beta-PR14000-237022d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13010/HighFidelity-Beta-Interface-PR14000-237022d.dmg

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13230/HighFidelity-Beta-PR14000-e15f22d.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13230/HighFidelity-Beta-PR14000-e15f22d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13230/HighFidelity-Beta-Interface-PR14000-e15f22d.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13230/HighFidelity-Beta-PR14000-e15f22d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13230/HighFidelity-Beta-Interface-PR14000-e15f22d.dmg

sabrina-shanman commented:
I've added a [link](https://github.com/sabrina-shanman/hifi/blob/stt_parenting2/docs/interface/raypick/pick-parenting.md.html) to the top post which documents the current pick behavior. I've also added a few additional caveats to my list of Caveats/Questions. These were some other issues that Sam was concerned about regarding the consistency of the pick parenting, specifically: Should parenting to a collision pick parent to its result like the other picks? And, should parenting to a ray/parabola/stylus pick affect the child's orientation?

sabrina-shanman commented:
I just merged in master, so the commit history should look a bit cleaner.

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13351/HighFidelity-Beta-PR14000-4011660.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13351/HighFidelity-Beta-Interface-PR14000-4011660.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13351/HighFidelity-Beta-PR14000-4011660.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13351/HighFidelity-Beta-Interface-PR14000-4011660.exe

sabrina-shanman commented:
Build this please

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13353/HighFidelity-Beta-PR14000-4011660.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13353/HighFidelity-Beta-PR14000-4011660.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13353/HighFidelity-Beta-Interface-PR14000-4011660.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13353/HighFidelity-Beta-PR14000-4011660.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13353/HighFidelity-Beta-Interface-PR14000-4011660.exe

SamGondelman commented on interface/src/Application.cpp:
```diff
@@ -2269,8 +2270,13 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
+        auto mouseRayPick = std::make_shared<RayPick>(glm::vec3(), Vectors::UP,
+            PickFilter(PickScriptingInterface::PICK_ENTITIES() | PickScriptingInterface::PICK_INCLUDE_NONCOLLIDABLE()), 0.0f, true);
+        mouseRayPick->parentTransform = std::make_shared<MouseTransformNode>();
+        auto mouseRayPickID = DependencyManager::get<PickManager>()->addPick(PickQuery::Ray, mouseRayPick);
```
This joint state needs to be set to the mouse, right?

SamGondelman commented on libraries/shared/src/RegisteredMetaTypes.h:
```diff
@@ -178,7 +178,9 @@ class PickRay : public MathPick {
         pickRay["origin"] = vec3toVariant(origin);
+        pickRay["position"] = vec3toVariant(origin);
         pickRay["direction"] = vec3toVariant(direction);
+        pickRay["orientation"] = quatToVariant(rotationBetween(Vectors::UP, direction));
```
I don't think we should change this.

sabrina-shanman commented on interface/src/raypick/ParabolaPick.cpp:
```diff
@@ -15,6 +15,43 @@
+PickParabola ParabolaPick::getMathematicalPick() const {
+    if (!parentTransform) {
+        PickParabola mathPick = _mathPick;
+        if (_rotateAccelerationWithAvatar) {
```
Unfortunately, since the parabola for the teleport is parented to the hand joint, gravity will be relative to the orientation of the joint, so _rotateAccelerationWithAvatar does serve a purpose.

... which reminds me, Sam mentioned the idea of having a separate parent transform for handling the acceleration to rotate with, as opposed to just the avatar. So, if you want to rotate acceleration relative to the avatar, you could pass in something like "accelerationParentID: MyAvatar.sessionUUID". Thoughts on this?

sabrina-shanman commented on interface/src/raypick/PickScriptingInterface.cpp:
```diff
@@ -373,6 +355,37 @@ void PickScriptingInterface::setPerFrameTimeBudget(unsigned int numUsecs) {
     DependencyManager::get<PickManager>()->setPerFrameTimeBudget(numUsecs);
 }
 
+PickQuery::JointState PickScriptingInterface::getPickJointState(const QVariantMap& propMap) {
```
Sure.

sabrina-shanman commented on interface/src/raypick/RayPick.cpp:
```diff
@@ -15,6 +15,17 @@
+    Transform currentParentTransform = parentTransform->getTransform();
+    Transform relativeTransform(rotationBetween(Vectors::UP, _mathPick.direction), glm::vec3(1.0f), _mathPick.origin);
+    Transform pickTransform = currentParentTransform.worldTransform(relativeTransform);
+    return PickRay(pickTransform.getTranslation(), pickTransform.getRotation() * Vectors::UP);
```
Yea, I really should just use currentParentTransform.transformDirection(_mathPick.direction).

SamGondelman commented:
To respond to your questions:
- the stylus pick thing is fine, but as you note, eventually we need to generalize styluses.
- The correct behavior with parenting to collision pick centers is fine, I think.  I think the correct solution is to make each collision point a joint, which we can do when we make Picks and Pointers proper SpatiallyNestables.
- I don't think we should use the normal for the pick transform.  I think it's fine to say Pick transforms just include position.  Eventually we'll have PointerParentTransforms that do contain rotation and scale.
- For both the rotate and scale with Avatar/Parent things, I think the best approach is: make them aliases, deprecate the "avatar" one, and remove soon.  Can you do a search through our scripts and make them use "parent"?
- I think all picks should scaleWithParent as the default, but the option would be good to have for all types.

sabrina-shanman commented:
Thank you for the feedback!

Regarding scaleWithAvatar... I was a bit hesitant to go through the scripts and change things because pointers also use it, but looking a bit deeper, it seems it's only used to determine the width of the pointer visualization, so well worth making it an alias in PointerScriptingInterface as well and making the script changes you mentioned.

Regarding rotateAccelerationWithAvatar: It can't be an alias because then the teleport script's parabola would have gravity relative to the orientation of the hand. We could have a separate transform node for rotating the acceleration, to make it more generalized.

Regarding scaleWithParent as the default: that sounds good then. I think scaleWithAvatar will have to have a different value from the default due to how it's used currently with parabola picks, but we could have the scaling as the default when using scaleWithParent going forward.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13451/HighFidelity-Beta-PR14000-8f2d3b1.apk

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13454/HighFidelity-Beta-PR14000-e51f48d.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13454/HighFidelity-Beta-PR14000-e51f48d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13454/HighFidelity-Beta-Interface-PR14000-e51f48d.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13454/HighFidelity-Beta-PR14000-e51f48d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13454/HighFidelity-Beta-Interface-PR14000-e51f48d.exe

sabrina-shanman commented:
The dismissal was from an accidental push which I have since reverted.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13464/HighFidelity-Beta-PR14000-e51f48d.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13464/HighFidelity-Beta-PR14000-e51f48d.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13464/HighFidelity-Beta-Interface-PR14000-e51f48d.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13464/HighFidelity-Beta-PR14000-e51f48d.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13464/HighFidelity-Beta-Interface-PR14000-e51f48d.exe

qlimnavarro commented:
Currently Testing

qlimnavarro commented:
Pass @conklin94122 
Automated Tests pass and teleport scales with Avatar. Performed quick smoke on Safe Teleport and it functions as expected even with different Avatar sizes. If more feedback is wanted, id be happy to assist.

sabrina-shanman commented:
I've discussed it with @samcake and we agreed it was minor enough for 74. It's also a continuation of previous work for pick parenting.

SamGondelman commented on interface/src/raypick/LaserPointer.cpp:
```diff
@@ -14,23 +14,24 @@
-            renderState->setLineWidth(lineWidth.toFloat());
+        QVariant pathWidth = pathProps.toMap()["lineWidth"];
+        if (pathWidth.isValid()) {
+            renderState->setPathWidth(pathWidth.toFloat());
```
Thank you for trying to consolidate this, but actually, I'd rather keep _pathWidth out of the base class.  Lasers and Parabolas both support "width", but future PathPointers might not.  Any path information needs to be kept in the derived RenderState classes.

SamGondelman commented on interface/src/raypick/ParabolaPick.cpp:
```diff
@@ -15,6 +15,45 @@
+        velocity = currentParentTransform.transformDirection(velocity);
+    } else {
+        glm::vec3 transformedVelocity = currentParentTransform.transformDirection(velocity);
+        velocity = glm::normalize(transformedVelocity) * glm::distance(velocity, Vectors::ZERO);
```
We should store `speed` in the constructor and use it here instead of glm::distance.

SamGondelman commented on interface/src/raypick/ParabolaPick.cpp:
```diff
@@ -15,6 +15,45 @@
+    }
+    if (_scaleWithParent) {
+        acceleration *= currentParentTransform.getScale();
+    }
```
This should be scaled before it's rotated, right?

sabrina-shanman replied:
The behavior was to rotate then scale, so I kept that behavior. I don't think it makes a difference in most cases, but for a non-uniform scaling situation, the current behavior is slightly more correct.

SamGondelman replied:
The old behavior only used a scalar for scale though, now it's a vec3. I think for a non-uniform situation, you would want to scale first, right?

sabrina-shanman replied:
Ah, you're right, that's how the transform works.

SamGondelman commented on interface/src/raypick/PathPointer.cpp:
```diff
@@ -146,12 +146,18 @@ void PathPointer::updateVisuals(const PickResultPointer& pickResult) {
     auto defaultRenderState = _defaultRenderStates.find(_currentRenderState);
+    float parentScale = 1;
+    if (_enabled && _scaleWithParent) {
+        glm::vec3 dimensions = DependencyManager::get<PickManager>()->getParentTransform(_pickUID).getScale();
```
Is this right?  The pick transform won't include scale.

SamGondelman commented on interface/src/raypick/PathPointer.cpp:
```diff
@@ -300,8 +304,8 @@ void StartEndRenderState::update(const glm::vec3& origin, const glm::vec3& end,
         if (distanceScaleEnd) {
             dim = getEndDim() * glm::distance(origin, end);
```
This should maybe also use `parentScale`?  I'm not sure.

SamGondelman replied:
What if you have something that's distanceScaleEnd and also scaleWithParent?  Right now it won't apply the parentScale, right?

sabrina-shanman replied:
Right. controllerDispatcher.js uses them both and scaleWithParent does nothing in those scripts.

I guess I can just change the scripts as well to just use distanceScaleEnd. I am a bit concerned that someone may have copied code from there, though.

sabrina-shanman replied:
Yea, I just tried it...

It reminded me that scaleWithParent is also used to scale the visualization of the path render state. 🤦‍♀️  So, it's actually not doing nothing. distanceScaleEnd is designed specifically with the avatar's laser pointer in mind, I think. Best left as-is.

SamGondelman commented on interface/src/raypick/PickScriptingInterface.cpp:
```diff
@@ -56,9 +52,9 @@ unsigned int PickScriptingInterface::createPick(const PickQuery::PickType type,
- * @property {string} [joint] Only for Joint or Mouse Ray Picks.  If "Mouse", it will create a Ray Pick that follows the system mouse, in desktop or HMD.
- *   If "Avatar", it will create a Joint Ray Pick that follows your avatar's head.  Otherwise, it will create a Joint Ray Pick that follows the given joint, if it
- *   exists on your current avatar.
+ * @property {Uuid} parentID - The ID of the parent, either an avatar, an entity, or an overlay.
```
Or pick?

SamGondelman commented on interface/src/raypick/PickScriptingInterface.cpp:
```diff
@@ -56,9 +52,9 @@ unsigned int PickScriptingInterface::createPick(const PickQuery::PickType type,
- *   If "Avatar", it will create a Joint Ray Pick that follows your avatar's head.  Otherwise, it will create a Joint Ray Pick that follows the given joint, if it
- *   exists on your current avatar.
+ * @property {Uuid} parentID - The ID of the parent, either an avatar, an entity, or an overlay.
+ * @property {number} parentJointIndex - The joint of the parent to parent to, for example, the joints on the model of an avatar. (default = 0, no joint)
```
You can denote defaults with `[parentJointIndex=0]`

SamGondelman commented on interface/src/raypick/PickScriptingInterface.cpp:
```diff
@@ -153,23 +140,26 @@ unsigned int PickScriptingInterface::createStylusPick(const QVariant& properties
+ * @property {boolean} [rotateAccelerationWithAvatar=true] Whether or not the acceleration axis should rotate with the avatar's local Y axis.
+ * @property {boolean} [rotateAccelerationWithParent=false] Whether or not the acceleration axis should rotate with the parent's local Y axis, if available.
+ * @property {boolean} [scaleWithParent=true] If true, the velocity and acceleration of the Pick will scale linearly with the parent, if available.
+ * @property {boolean} [scaleWithAvatar] Alias for scaleWithParent. <em>Deprecated.</em>
```
Don't list this as a separate property.  Put it on the same line as above (`"scaleWithAvatar" is an alias but is deprecated"`)

SamGondelman commented on interface/src/raypick/PickScriptingInterface.cpp:
```diff
@@ -373,51 +358,63 @@ void PickScriptingInterface::setPerFrameTimeBudget(unsigned int numUsecs) {
-                }
-            }
-        }
+void PickScriptingInterface::setParentTransform(std::shared_ptr<PickQuery> pick, const QVariantMap& propMap) {
```
Have this take in a `const std::shared_ptr<PickQuery>&`

sabrina-shanman replied:
In this scenario pick is modified. I initially tried "std::shared_ptr<PickQuery>&" instead, but the compiler really didn't like it for some reason.

SamGondelman replied:
Whoops good point

SamGondelman commented on interface/src/raypick/PointerScriptingInterface.cpp:
```diff
@@ -97,7 +97,8 @@ unsigned int PointerScriptingInterface::createStylus(const QVariant& properties)
  * @property {boolean} [distanceScaleEnd=false] If true, the dimensions of the end of the Pointer will scale linearly with distance.
- * @property {boolean} [scaleWithAvatar=false] If true, the width of the Pointer's path will scale linearly with your avatar's scale.
+ * @property {boolean} [scaleWithParent=false] If true, the width of the Pointer's path will scale linearly with the pick parent's scale.
+ * @property {boolean} [scaleWithAvatar] Alias for scaleWithParent
```
Move to above line and mark deprecated.

SamGondelman commented on interface/src/raypick/PointerScriptingInterface.cpp:
```diff
@@ -248,7 +251,8 @@ unsigned int PointerScriptingInterface::createLaserPointer(const QVariant& prope
 * @property {boolean} [distanceScaleEnd=false] If true, the dimensions of the end of the Pointer will scale linearly with distance.
-* @property {boolean} [scaleWithAvatar=false] If true, the width of the Pointer's path will scale linearly with your avatar's scale.
+* @property {boolean} [scaleWithParent=true] If true, the width of the Pointer's path will scale linearly with the pick parent's scale.
+* @property {boolean} [scaleWithAvatar] Alias for scaleWithParent
```
Move above.

SamGondelman commented on interface/src/raypick/RayPick.h:
```diff
@@ -70,7 +70,15 @@ class RayPickResult : public PickResult {
 public:
-    RayPick(const PickFilter& filter, float maxDistance, bool enabled) : Pick(filter, maxDistance, enabled) {}
+    RayPick(const PickFilter& filter, float maxDistance, bool enabled) :
+        Pick(PickRay(Vectors::ZERO, Vectors::UP), filter, maxDistance, enabled) {
```
Just use `PickRay()`

SamGondelman commented on libraries/shared/src/RegisteredMetaTypes.h:
```diff
@@ -198,6 +198,8 @@ void pickRayFromScriptValue(const QScriptValue& object, PickRay& pickRay);
 class StylusTip : public MathPick {
 public:
     StylusTip() : position(NAN), velocity(NAN) {}
+    StylusTip(bilateral::Side side, glm::vec3 position = Vectors::ZERO, glm::quat orientation = Quaternions::IDENTITY, glm::vec3 velocity = Vectors::ZERO) :
```
Pass these things by const reference

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13499/HighFidelity-Beta-PR14000-0cf7e95.apk

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13502/HighFidelity-Beta-PR14000-7a89f95.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13502/HighFidelity-Beta-PR14000-7a89f95.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13502/HighFidelity-Beta-Interface-PR14000-7a89f95.dmg

SamGondelman commented on interface/src/raypick/RayPick.h:
```diff
@@ -70,7 +70,15 @@ class RayPickResult : public PickResult {
 public:
-    RayPick(const PickFilter& filter, float maxDistance, bool enabled) : Pick(filter, maxDistance, enabled) {}
+    RayPick(const PickFilter& filter, float maxDistance, bool enabled) :
+        Pick(PickRay(), filter, maxDistance, enabled) {
```
I was thinking about this more and I think I was wrong when I asked you to add this.  I realized origin and direction need to be defined.  You can remove this constructor and put the thing in Application.cpp back how you had it.

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13502/HighFidelity-Beta-PR14000-7a89f95.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13502/HighFidelity-Beta-Interface-PR14000-7a89f95.exe

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13510/HighFidelity-Beta-PR14000-b988382.apk

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13510/HighFidelity-Beta-PR14000-b988382.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13510/HighFidelity-Beta-Interface-PR14000-b988382.exe

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13510/HighFidelity-Beta-PR14000-b988382.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13510/HighFidelity-Beta-Interface-PR14000-b988382.dmg

qaAdamSmith82 commented:
Currently Testing

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13543/HighFidelity-Beta-PR14000-67dc526.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13543/HighFidelity-Beta-PR14000-67dc526.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13543/HighFidelity-Beta-Interface-PR14000-67dc526.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13543/HighFidelity-Beta-PR14000-67dc526.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13543/HighFidelity-Beta-Interface-PR14000-67dc526.exe

conklin94122 commented:
This isn't on the approved list for 74.0. I'm changing the milestone to 75.0. Please discuss with Brad.

sabrina-shanman commented:
Re-merged master due to conflict with Samuel's new drawInFront parabola feature.

hifi-gustavo commented:
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/13608/HighFidelity-Beta-PR14000-2ef85df.apk

hifi-gustavo commented:
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13608/HighFidelity-Beta-PR14000-2ef85df.dmg

hifi-gustavo commented:
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/13608/HighFidelity-Beta-Interface-PR14000-2ef85df.dmg

hifi-gustavo commented:
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13608/HighFidelity-Beta-PR14000-2ef85df.exe

hifi-gustavo commented:
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/13608/HighFidelity-Beta-Interface-PR14000-2ef85df.exe

divyaelango88 commented:
@SamGondelman @samcake  Alisa wants to make sure we run the safe teleport target tests thoroughly for this PR.,.. since she is raising this as risky, we might also want to consider pushing this to 75.

qlimnavarro commented:
Currently Testing

qlimnavarro commented:
Pass @conklin94122 

Tests pass and teleport scales. Performed quick smoke on Safe Teleport and it functions as expected even with different Avatar sizes.

conklin94122 commented:
To be clear: this is NOT on the scheduled list of changes for 74.0. If Brad clears it to go in, it can be merged. Until then it will not go into 74.0

divyaelango88 commented:
Chatting with Brad about this right now. @samcake really wants this in.
